### PR TITLE
Add new certificate script commands to facilitate DPS and manual X.509 flows

### PR DESCRIPTION
**Pull request by mrohera**
The changes introduce new commands that are very explicit about what type of certificate is being created. Either the gateway CA certificate or and identity certificate.

### CODE REVIEW

File: tools/CACertificates/ca-certs.ps1 [hunk 1]
```diff
@@ -750,12 +755,58 @@ function New-CACertsDevice([Parameter(Mandatory=$TRUE)][string]$deviceName, [str
         CA certificate is used.
 #>
 function New-CACertsEdgeDevice([Parameter(Mandatory=$TRUE)][string]$deviceName, [string]$issuerPrefix=$_intermediatePrefix)
+{
+    if ([string]::IsNullOrWhiteSpace($deviceName))
```
Comment (damonbarry) [inline tools/CACertificates/ca-certs.ps1]: Because `$deviceName` is a mandatory parameter, it already can't be null or empty. So that only leaves whitespace, which could be handled with the attribute `[ValidatePattern("[^\s]+")]`.

File: tools/CACertificates/ca-certs.ps1 [hunk 2]
```diff
@@ -750,12 +755,58 @@ function New-CACertsDevice([Parameter(Mandatory=$TRUE)][string]$deviceName, [str
...
+    New-CACertsDevice $deviceName $issuerPrefix "iot-edge-device"
+}
+
+<#
+    .SYNOPSIS
+    Generate an IoT Edge device CA certificate, private key and full certificate chain
+    using an issuer's certificate and key
+    .DESCRIPTION
+    Generate a certificate using the supplied common name and have it signed by
+    an issuer certificate and key. Issuer is spcified by its prefix and its corresponding
```
Comment (damonbarry) [inline tools/CACertificates/ca-certs.ps1]: typo: spcified

Comment (damonbarry) [reply tools/CACertificates/ca-certs.ps1]: (Same for the next function description)

File: tools/CACertificates/certGen.sh [hunk 3]
```diff
@@ -367,7 +367,7 @@ function generate_verification_certificate()
 ###############################################################################
 # Generates a certificate for a device, chained to the intermediate.
 ###############################################################################
-function generate_device_certificate()
+function generate_identity_device_certificate()
```
Comment (damonbarry) [inline tools/CACertificates/certGen.sh]: Should this function by `generate_device_identity_certificate`?

File: tools/CACertificates/certGen.sh [hunk 4]
```diff
@@ -404,16 +423,30 @@ function generate_edge_server_certificate()
 }
 
 ###############################################################################
-# Generates a certificate for a Edge device, chained to the intermediate.
+# Generates a CA certificate for a Edge device, chained to the intermediate.
 ###############################################################################
-function generate_edge_device_certificate()
+function generate_edge_device_ca_certificate()
```
Comment (damonbarry) [inline tools/CACertificates/certGen.sh]: Why did you remove parameter checking on this function? I ask because the previous function (`generate_edge_device_identity_certificate`) has parameter checking...

File: tools/CACertificates/ca-certs.ps1 [hunk 5]
```diff
@@ -708,18 +708,24 @@ function New-CACertsVerificationCert([Parameter(Mandatory=$TRUE)][string]$reques
     .PARAMETER isEdgeDevice
```
Comment (damonbarry) [inline tools/CACertificates/ca-certs.ps1]: The help text needs to be updated with the new `filePrefix` parameter that replaces `isEdgeDevice`.

File: tools/CACertificates/README.md [hunk 6]
```diff
@@ -78,22 +78,30 @@ On Azure IoT Hub, navigate to the "Device Explorer".  Add a new device (e.g. `my
 Note that if you're using this certificate as a DPS registration ID, the ID **must be lower case** or the server will reject it.
 
 ### **PowerShell**
-#### IoT Leaf Device
-* Run `New-CACertsDevice mydevice` to create the new device certificate.
```
Comment (damonbarry) [inline tools/CACertificates/README.md]: Both old and new end-to-end tests call `New-CACertsDevice` (and `./certGen.sh create_edge_device_certificate`). It looks like those will continue to work, but should the tests be updated to use the "recommended" approach?

File: tools/CACertificates/ca-certs.ps1 (hunk 1 above)
Comment (mrohera) [reply tools/CACertificates/ca-certs.ps1]: I prefer a friendly error message even though the null check is redundant.

File: tools/CACertificates/certGen.sh (hunk 3 above)
Comment (mrohera) [reply tools/CACertificates/certGen.sh]: Agreed and fixed.

File: tools/CACertificates/certGen.sh (hunk 4 above)
Comment (mrohera) [reply tools/CACertificates/certGen.sh]: The validations were not put in here and other the function. I have gone ahead and added these. Thanks.

File: tools/CACertificates/README.md (hunk 6 above)
Comment (mrohera) [reply tools/CACertificates/README.md]: Correct. Longer term the E2E tests should be migrated over to the newer commands and the older one can simply be removed.

Comment (damonbarry) [reply tools/CACertificates/README.md]: Agree it should be done separately.
